### ISSUE TITLE
VersionInfo.parse should be a class method

### ISSUE THREAD

**Issue by Anvil** [python-semver/python-semver#276]
I wanted to create a subclass of VersionInfo for company reasons (more checkings, more methods, etc) and it turned out the VersionInfo.parse is a staticmethod, always returning VersionInfo instances, no matter what.

By turning the `parse` static method into a class method and making it `return cls(**version_parts)` the result of parse will be an instance of the subclass (my class), not the superclass (VersionInfo).

e.g:

```python

class MyBogusVersionInfo(semver.VersionInfo):

    @classmethod
    def parse(cls, version):
        if version.startswith("foobar"):
            version = version[6:]
        return super().parse(version)

print(type(MyBogusVersionInfo.parse("1.2.3")))
```

Current result is : `<class 'semver.VersionInfo'>`, while I would have expected `<class 'MyBogusVersionInfo'>`

**Comment by tomschr** [comment]
Thanks @Anvil for your issue and the pull request. :+1: 

That's an interesting approach. I see your point. However, not entirely sure if changing `.parse` from a staticmethod to a classmethod is completely backward compatible... :thinking: 

Apart from that, the `.parse` method could be removed in semver3. We've discussed this in issue #258. For semver3, an idea is to "overload" the initializer to allow more datatypes. Maybe that could solve your use case. If we go this route, that would make the `.parse` and adapting it kind of obsolete.

Do you need it for Python2 only or do you use it for Python3?

Thanks for your pull request (PR). As the above discussion is not really solved, I would like to wait for this issue until the above issue is remedied.

Just a short comment about your PR: you don't have to change anything right now (due to the open discussion). It's just a reminder for future PRs. We require test cases for each new feature. :wink: If a feature changes the public API, this also needs to be documented. Ideally with some code examples. For further details, refer to our [Contributing to semver](https://python-semver.readthedocs.io/en/latest/development.html) section.

**Comment by Anvil** [comment]
Hello

As far as I can tell, there wont be any breaking by changing static to class. The reference "VersionInfo.parse" makes it automatically bound to the class.

In general, *all* factory methods (methods that return an instance of the class) should be class methods, not staticmethods.

I dont use python 2 anymore. Do you intent do keep compatibility with python2 ?

**Comment by Anvil** [comment]
I can provide tests and examples if it's accepted, though. Feel free to ask.

**Comment by Anvil** [comment]
A basic yet (i think) meaningful test case / example, would be to have a subclass that accepts "v1.2.3" as semver, so that you can directly feed git tag ot the parse method. The subclass would  have a parse such as:

```python
class GitTagSemVer(VersionInfo):
    @classmethod
    def parse(cls, version: str):
        if not version.startswith('v'):
            raise ValueError(f"{version}: not a valid semantic version tag") 
        return super().parse(version[1:])

    def __str__(self):
        # Reconstruct the tag.
        return "v" + super().__str__()

import git
import logging

def semver_tags(path):
    repo = git.Repo(path)
    for tag in repo.tags:
        try:
            yield GitTagSemVer(tag)
        except ValueError:
            logging.error(f"{tag}: bad semver tag")
```

**Comment by tomschr** [comment]
Thank you @Anvil for all the ideas. Sorry for not responding to it earlier. I hope I can work on that soon.

> Do you intent do keep compatibility with python2 ?

That's a good question. If possible, I would love to have at least this issue compatible between the two versions. It doesn't look like a big problem. But let me think of it again, once I'm coming back to this issue.

**Comment by Anvil** [comment]
I'm glade it makes sense. On my part, there is no requirement nor need for a python2 compatibility. My team and I are working with python >= 3.6. Thank you for your work !

**Comment by tomschr** [comment]
> I'm glade it makes sense. 

Absolutely! :+1:  I'm glad you've reached out to us and contributed. 

> On my part, there is no requirement nor need for a python2 compatibility. My team and I are working with python >= 3.6. Thank you for your work !

That's great! I've added a small test case; there I had to adapt the `super()` method a bit. I almost forgot how that worked as I'm using Python 3 most of the day.

Plan is to release 2.11.0 soon and start working on Python 3. Otherwise these compatibility thing always bites me. :laughing:
